### PR TITLE
[CUMULUS-755] DynamoDB autoscaling

### PR DESCRIPTION
**Pull request by abarciauskas-bgse**
**Summary:** Add support to @cumulus/deployment and @cumulus/api for autoscaling DynamoDB

Addresses [CUMULUS-755: Add DynamoDB autoscaling support](https://bugs.earthdata.nasa.gov/browse/CUMULUS-755)

## Changes

- Values defined in `app/config.yml` are used to replace values `@cumulus/api`'s `tables.yml`.
- Values in `@cumulus/api`'s `tables.yml` are used to template (via handlebars) the final application's `app/cloudformation.yml` via `@cumulus/deployment`'s `cloudformation.template.yml`

## Test Plan

- [x] Unit tests
- [x] Adhoc testing
- [x] Update CHANGELOG

Also tested functionality detailed on [this nasa wiki page](https://wiki.earthdata.nasa.gov/pages/editpage.action?pageId=112496715)

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -5,12 +5,16 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
...
 
 ## [Unreleased]
-### Added
-- **CUMULUS-491** - Add granule reconciliation API endpoints.
 
 ### Added
+- **CUMULUS-491** - Add granule reconciliation API endpoints.
 - **CUMULUS-508** - `@cumulus/deployment` cloudformation template allows for lambdas and ECS clusters to have multiple AZ availability.
     - `@cumulus/deployment` also ensures docker uses `devicemapper` storage driver.
+- **CUMULUS-755** - Add DynamoDB autoscaling support
```
Comment (laurenfrederick) [inline CHANGELOG.md]: Seems like we will need some documentation on this on cumulus-nasa.github.io

File: example/app/config.yml [hunk 2]
```diff
@@ -86,6 +86,7 @@ default:
     stepRoleArn: 'arn:aws:iam::{{AWS_ACCOUNT_ID}}:role/{{prefix}}-integration-steprole'
     instanceProfile: 'arn:aws:iam::{{AWS_ACCOUNT_ID}}:instance-profile/{{prefix}}-integration-ecs'
     distributionRoleArn: 'arn:aws:iam::{{AWS_ACCOUNT_ID}}:role/{{prefix}}-integration-distribution-api-lambda'
+    scalingRoleArn: 'arn:aws:iam::{{AWS_ACCOUNT_ID}}:role/{{prefix}}-integration-scaling-role'
```
Comment (bilts) [inline example/app/config.yml]: Need to get these roles into the PO.DAAC NGAP 2.0 instance and LPDAAC NGAP 1.x instance

File: packages/deployment/iam/cloudformation.template.yml [hunk 3]
```diff
@@ -481,6 +481,34 @@ Resources:
                   - s3:ListAllMyBuckets
                 Resource: '*'
 
+  ScalingRole:
```
Comment (laurenfrederick) [inline packages/deployment/iam/cloudformation.template.yml]: Why do we need a separate role for scaling?

File: packages/deployment/app/config.yml [hunk 4]
```diff
@@ -95,6 +95,70 @@ default:
 
   # set to true to activate pointInTime on all tables
   enablePointInTime: false
+  tables:
+    # REVIEW: Autoscaling policy configuration is the same for all tables, to
```
Comment (bilts) [inline packages/deployment/app/config.yml]: Is there a way that we could use some sort of sensible default for all tables and then override one-by-one?  Seems like a ton of boilerplate to say the same thing over and over and makes adding a new table more cumbersome.

File: packages/deployment/iam/cloudformation.template.yml (hunk 3 above)
Comment (bilts) [inline packages/deployment/iam/cloudformation.template.yml]: (Mentioned above) Needs to make it into NGAP environments.  Do we have a process for getting that in place?

File: packages/api/config/tables.yml [hunk 5]
```diff
@@ -2,6 +2,19 @@ CollectionsTable:
   read: 5
   write: 1
   pointInTime: '{{enablePointInTime}}'
+  enableAutoScaling: '{{tables.collections.enableAutoScaling}}'
```
Comment (bilts) [inline packages/api/config/tables.yml]: It seems like we're specifying all this info in 2 places and I don't understand the value of doing so

File: packages/api/config/tables.yml [hunk 6]
```diff
@@ -2,6 +2,19 @@ CollectionsTable:
...
+  autoScaling:
+    roleArn: '{{iams.scalingRoleArn}}'
+    readCapacityUnits:
+      min: '{{tables.collections.readCapacityUnits.min}}'
```
Comment (bilts) [inline packages/api/config/tables.yml]: Wherever this ends up (again, I think it should be condensed into a single doc), consider something like:

    min: {{tables.collections.readCapacityUnits.min || something.autoscaling.readCapacityUnitDefaults.min}}

That way we can set sensible global defaults and adjust as needed.

Comment (bilts) [reply packages/api/config/tables.yml]: Obviously similar for all min/max fields and anything that gets repeated over and over with the same values in almost every location.

File: CHANGELOG.md (hunk 1 above)
Comment (abarciauskas-bgse) [reply CHANGELOG.md]: Agree, will open a PR

File: packages/deployment/iam/cloudformation.template.yml (hunk 3 above)
Comment (abarciauskas-bgse) [reply packages/deployment/iam/cloudformation.template.yml]: This is a role for DynamoDB to make changes based on current througput. Application Autoscaling (what DynamoDB uses) relies on cloudwatch alarms to trigger scaling.

This is detailed in the AWS Blog: [How to use AWS CloudFormation to configure auto scaling for Amazon DynamoDB tables and indexes](https://aws.amazon.com/blogs/database/how-to-use-aws-cloudformation-to-configure-auto-scaling-for-amazon-dynamodb-tables-and-indexes/)

and you can also see DynamoDB console requires a role in the console:
screen shot 2018-06-28 at 6 13 54 pm

File: packages/api/config/tables.yml (hunk 5 above)
Comment (abarciauskas-bgse) [reply packages/api/config/tables.yml]: An application owner can have control over which tables use auto-scaling by configuring these values in their deployment's `app/config.yml` file

File: CHANGELOG.md (hunk 1 above)
Comment (abarciauskas-bgse) [reply CHANGELOG.md]: Although, I'm going to wait on opening a PR until I'm certain we are going with a given implementation for configuring the autoscaling options

File: packages/api/config/tables.yml (hunk 5 above)
Comment (bilts) [reply packages/api/config/tables.yml]: The comment wasn't just for this particular line but for 5-17.  All of those lines are just referencing variables from more or less one place (could easily be made into 1 place by adding indirection to `iams.scalingRoleArn`).

I don't understand why all the tables each take 13 lines to say "Use the configuration values over there...".  Could we just provide a way to say that?

Comment (bilts) [reply packages/api/config/tables.yml]: It probably seems nitpicky, but more lines = more errors and more clutter that makes it harder to spot what's actually going on.

**Pull request merged**